### PR TITLE
[Delivers #143245661] Hugsql substitution for couch

### PR DESCRIPTION
**Pull request by dguettler**
## TODO

- [x] timestamps format `2017-12-06T04:29:54.000` but was `2016-01-26T16:12:55.280Z`
- [x] updated-at currently added when running [doc|value]-to-db but not updated in doc or value, change order when updating, apply updates then select entity from db and return
- [x] posting a note from UI results in `Data truncation: Incorrect datetime value: '2017-12-13T04:30:12.395Z' for column 'timestamp' at row 1`
- [x] Adding a tag to a source - Missing :project_id
- [x] Update unit tests
- [x] Update source migration to rollback if any of the projects is missing

## UI
- [x] Activities UI "Create an activity" missing action handler
- [x] Activity UI - Add operator, when selecting the selection is invisible in the dropdown
- [x] Create source call successful, but source is not displayed. Present after page reload
- [x] Adding tag to source fails - `Parameter Mismatch: :project_id parameter data not found.`
- [x] Adding a remote file fails - `Parameter Mismatch: :resource_id parameter data not found.`
- [x] Note UI - Text is overflowing
- [x] Adding a member to a project results in 404, seems like something is not quite working with Rails <> API exchange here
- [x] Loading screen for groups/:id is wrong
- [x] Load relations for sources so we can show it exists without having access to the source

### Activities
- [X] GET /api/v1/o/{org}/activities
- [ ] POST /api/v1/o/{org}/activities - failing on prod and dev
- [X] DELETE /api/v1/o/{org}/activities/{id}
- [X] GET /api/v1/o/{org}/activities/{id}
- [X] PUT /api/v1/o/{org}/activities/{id}
- [x] GET /api/v1/o/{org}/activities/{id}/links/{rel}
- [x] GET /api/v1/o/{org}/activities/{id}/links/{rel}/relationships
- [X] POST /api/v1/o/{org}/activities/{id}/links/{rel}/relationships

### Admin
- [X] GET /api/v1/o
- [ ] POST /api/v1/o - Failing on the rails side (recurly)
- [X] DELETE /api/v1/{id}
- [X] GET /api/v1/{id}
- [ ] PUT /api/v1/{id} - Currently allows to update org 0, maybe because I'm an admin? Same behavior on production
- [x] GET /api/v1/o/{org}/stats
- [X] GET /api/v1/o/{org}/memberships
- [X] POST /api/v1/o/{org}/memberships
- [X] DELETE /api/v1/o/{org}/memberships/{id}
- [X] PUT /api/v1/o/{org}/memberships/{id}
- [X] GET /api/v1/o/{org}/groups
- [x] POST /api/v1/o/{org}/groups
- [x] DELETE /api/v1/o/{org}/groups/{id}
- [X] GET /api/v1/o/{org}/groups/{id}
- [X] PUT /api/v1/o/{org}/groups/{id}
- [X] GET /api/v1/o/{org}/groups/{id}/memberships
- [x] POST /api/v1/o/{org}/groups/{id}/memberships - Allows invalid membership_id on the Rails side
- [x] POST /api/v1/o/{org}/groups/{id}/memberships/{id}
- [X] GET /api/v1/o/{org}/groups/{id}/memberships/{id}
- [ ] PUT /api/v1/o/{org}/groups/{id}/memberships/{id} - Unsupported on the Rails side

### Annotations
- [X] GET /api/v1/o/{org}/annotations/{id} - Broken in production (disallowed-key :edited_at)
- [x] GET /api/v1/o/{org}/entities/{id}/annotations/tags
- [X] POST /api/v1/o/{org}/entities/{id}/annotations/tags
- [x] DELETE /api/v1/o/{org}/entities/{id}/annotations/tags/{id}
- [x] GET /api/v1/o/{org}/entities/{id}/annotations/properties
- [X] POST /api/v1/o/{org}/entities/{id}/annotations/properties
- [x] DELETE /api/v1/o/{org}/entities/{id}/annotations/properties/{id}
- [x] GET /api/v1/o/{org}/entities/{id}/annotations/notes
- [X] POST /api/v1/o/{org}/entities/{id}/annotations/notes
- [x] DELETE /api/v1/o/{org}/entities/{id}/annotations/notes/{id}
- [X] PUT /api/v1/o/{org}/entities/{id}/annotations/notes/{id}

### Auth
- [X] GET /api/v1/o/{org}/authorizations

### Default
- [X] GET /healthz

### Entities
- [X] DELETE /api/v1/o/{org}/entities/{id}
- [X] GET /api/v1/o/{org}/entities/{id}
- [X] PUT /api/v1/o/{org}/entities/{id}/restore

### Files
- [X] GET /api/v1/o/{org}/files
- [x] DELETE /api/v1/o/{org}/files/{id}
- [X] GET /api/v1/o/{org}/files/{id}
- [X] POST /api/v1/o/{org}/files/{id}
- [X] PUT /api/v1/o/{org}/files/{id}
- [x] POST /api/v1/o/{org}/files/{id}/move
- [x] GET /api/v1/o/{org}/files/{id}/heads
- [X] GET /api/v1/o/{org}/files/{id}/links/{rel}
- [X] GET /api/v1/o/{org}/files/{id}/links/{rel}/relationships
- [X] POST /api/v1/o/{org}/files/{id}/links/{rel}/relationships
- [X] DELETE /api/v1/o/{org}/revisions/{id}
- [X] GET /api/v1/o/{org}/revisions/{id}
- [X] POST /api/v1/o/{org}/revisions/{id}
- [X] PUT /api/v1/o/{org}/revisions/{id}
- [ ] PUT /api/v1/o/{org}/revisions/{id}/upload-complete - Fails on pubsub
- [X] PUT /api/v1/o/{org}/revisions/{id}/upload-failed
- [X] GET /api/v1/o/{org}/revisions/{id}/links/{rel}
- [X] GET /api/v1/o/{org}/revisions/{id}/links/{rel}/relationships
- [x] POST /api/v1/o/{org}/revisions/{id}/links/{rel}/relationships

### Folders
- [X] GET /api/v1/o/{org}/folders
- [X] DELETE /api/v1/o/{org}/folders/{id}
- [X] GET /api/v1/o/{org}/folders/{id}
- [x] POST /api/v1/o/{org}/folders/{id}
- [X] PUT /api/v1/o/{org}/folders/{id}
- [x] POST /api/v1/o/{org}/folders/{id}/move
- [X] GET /api/v1/o/{org}/folders/{id}/links/{rel}
- [X] GET /api/v1/o/{org}/folders/{id}/links/{rel}/relationships
- [X] POST /api/v1/o/{org}/folders/{id}/links/{rel}/relationships

### Links
- [X] DELETE /api/v1/o/{org}/relationships/{id}
- [X] GET /api/v1/o/{org}/relationships/{id}

### Projects
- [X] GET /api/v1/o/{org}/projects
- [x] POST /api/v1/o/{org}/projects
- [X] DELETE /api/v1/o/{org}/projects/{id}
- [X] GET /api/v1/o/{org}/projects/{id}
- [X] POST /api/v1/o/{org}/projects/{id}
- [X] PUT /api/v1/o/{org}/projects/{id}
- [X] GET /api/v1/o/{org}/projects/{id}/links/{rel}
- [X] GET /api/v1/o/{org}/projects/{id}/links/{rel}/relationships
- [X] POST /api/v1/o/{org}/projects/{id}/links/{rel}/relationships

### Provenance
- [X] GET /api/v1/o/{org}/prov/{id}

### Search
- [ ] GET /api/v1/o/{org}/search - Connection refused

### Sources
- [X] GET /api/v1/o/{org}/sources
- [X] POST /api/v1/o/{org}/sources
- [X] DELETE /api/v1/o/{org}/sources/{id}
- [X] GET /api/v1/o/{org}/sources/{id}
- [X] POST /api/v1/o/{org}/sources/{id} - Error entities: [ (not (some-matching-either-clause? a-clojure.lang.PersistentArrayMap)) ]
- [X] PUT /api/v1/o/{org}/sources/{id}
- [x] GET /api/v1/o/{org}/sources/{id}/links/{rel}
- [x] GET /api/v1/o/{org}/sources/{id}/links/{rel}/relationships
- [x] POST /api/v1/o/{org}/sources/{id}/links/{rel}/relationships

### Teams
- [X] GET /api/v1/o/{org}/teams/{id}
- [X] GET /api/v1/o/{org}/teams/{id}/team_groups - ID must be INT
- [X] POST /api/v1/o/{org}/teams/{id}/team_groups - Should there be some permissions check?
- [X] DELETE /api/v1/o/{org}/teams/{id}/team_groups/{gid}
- [X] GET /api/v1/o/{org}/teams/{id}/team_groups/{gid}
- [X] PUT /api/v1/o/{org}/teams/{id}/team_groups/{gid}
- [X] POST /api/v1/o/{org}/teams/{id}/memberships - Need to disable Cloudant on the Rails side
- [X] DELETE /api/v1/o/{org}/teams/{id}/memberships/{mid} - Need to disabled Cloudant on the Rails side
- [X] PUT /api/v1/o/{org}/teams/{id}/memberships/{mid}
- [X] GET /api/v1/o/{org}/roles

### UI
- [X] GET /api/v1/o/{org}/breadcrumbs
- [X] POST /api/v1/o/{org}/breadcrumbs

### Zip
- [X] GET /api/v1/o/{org}/zip/folders/{id} - Sends 307 redirect
- [X] GET /api/v1/o/{org}/zip/activities/{id} - Sends 307 redirect

### CODE REVIEW

File: src/ovation/core.clj [hunk 1]
```diff
@@ -1,122 +1,359 @@
...
+  (let [msg-fn   (fn [doc]
+                   {:id           (str (:_id doc))
+                    :type         (:type doc)
+                    :organization (:organization_id doc)})
+        topic    (config/config :db-updates-topic :default :updates)
+        channels (map #(pubsub/publish publisher topic (msg-fn %) (chan)) docs)]
+
+    (async/pipe (async/merge channels) channel close?)))
+
+(defn publish-updates
```
Comment (dguettler) [inline src/ovation/core.clj]: @barrywark placeholder still need to actually call this for updates / inserts.
But not entirely sure how this should look like

File: src/ovation/core.clj [hunk 2]
```diff
@@ -1,122 +1,359 @@
...
+        args {:id id,
+              :archived include-trashed
+              :organization_id org-id}]
+    (-> (or (notes/find-by-uuid db args)
+            (properties/find-by-uuid db args)
+            (tags/find-by-uuid db args)
+            (timeline_events/find-by-uuid db args))
+      (tr/entity-from-db ctx))))
+
+(defn -get-entities
```
Comment (barrywark) [inline src/ovation/core.clj]: Could you consolidate this with get-entities-by-id? Looks like same shape code. Could you pass `[]` for the ids?

File: src/ovation/core.clj [hunk 3]
```diff
@@ -1,122 +1,359 @@
...
-        k/PROJECT-TYPE [(:_id doc)]
+        c/PROJECT-TYPE [(:_id doc)]
         ;; default
-        (-> doc
-          :links
-          :_collaboration_roots))
+        [(:project_id doc)])
       [])))
 
+(defn --create-entity
```
Comment (barrywark) [inline src/ovation/core.clj]: I might call this `make-entity` since it's just constructing a map

Comment (barrywark) [reply src/ovation/core.clj]: nope. nvm.

**Comment by dguettler** [comment]
@barrywark take a look at https://github.com/physion/ovation-webapi/pull/362/commits/d32639ea120852e742438a3c011cf8b8514d961e those are the changes to associated a Source with a Project

**Comment by barrywark** [comment]
@dguettler should this PR be closed?

**Comment by barrywark** [comment]
@dguettler can be closed, right?

**Pull request closed without merging**